pkg/app/piped/trigger/determiner.go

```diff
@@ -61,14 +63,41 @@ func (d *OnCommandDeterminer) ShouldTrigger(_ context.Context, _ *model.Applicat
 }
 
 type OnOutOfSyncDeterminer struct {
+	client apiClient
 }
 
 // ShouldTrigger decides whether a given application should be triggered or not.
-func (d *OnOutOfSyncDeterminer) ShouldTrigger(_ context.Context, _ *model.Application, appCfg *config.GenericDeploymentSpec) (bool, error) {
+func (d *OnOutOfSyncDeterminer) ShouldTrigger(ctx context.Context, app *model.Application, appCfg *config.GenericDeploymentSpec) (bool, error) {
 	if *appCfg.Trigger.OnOutOfSync.Disabled {
 		return false, nil
 	}
 
+	// Find the most recently triggered deployment.
+	// Nil means it seems the application has been added recently
+	// and no deployment was triggered yet.
+	ref := app.MostRecentlyTriggeredDeployment
+	if ref == nil {
+		return true, nil
+	}
+
+	resp, err := d.client.GetDeployment(ctx, &pipedservice.GetDeploymentRequest{
+		Id: ref.DeploymentId,
+	})
+	if err != nil {
+		return false, err
+	}
+	deployment := resp.Deployment
+
+	// Check if it was already completed or not.
+	if !model.IsCompletedDeployment(deployment.Status) {
+		return false, nil
+	}
+
+	// Check the elapsed time since the last deployment.
+	if time.Since(time.Unix(deployment.CompletedAt, 0)) < appCfg.Trigger.OnOutOfSync.MinWindow.Duration() {
+		return false, nil
+	}
+
```

Comment:
tbh, it looks like this determiner will trigger a new deployment whenever
- no previous triggered deployment (newly added application).
- time since that latest triggered deployment overs a `minWindow` of time.

Those two are good to me, but since this determiner is for the `OUT_OF_SYNC` state, I think we may need logic to trigger a new deployment on `app.SyncState.Status` is `OUT_OF_SYNC`. Please correct me if I'm wrong.

Reply:
> it looks like this determiner will trigger a new deployment whenever
> - no previous triggered deployment (newly added application).
> - time since that latest triggered deployment overs a minWindow of time.

This is correct.

> I think we may need logic to trigger a new deployment on app.SyncState.Status is OUT_OF_SYNC. 

We don't need that logic because this determiner is accepting only OUT_OF_SYNC candidates.
https://github.com/pipe-cd/pipe/blob/c00e51ac61a93399ccc5e572468c1e23d28d9e8f/pkg/app/piped/trigger/trigger.go#L312
https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/trigger/trigger.go#L237